File: XCamp/src/CS401H/CS401H_TreeMatching.java
package CS401H;
import java.util.*;
import java.io.*;

public class CS401H_TreeMatching {
	
	static int n;
	static final int M = 2*(int)10e5+1;
	static ArrayList<Integer>[] adj;
	static int[][] dp = new int[M][2];
	static BufferedReader in;
	static StringTokenizer st;
	
	public static void main(String[] args) throws IOException {
		in = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(in.readLine());
		adj = new ArrayList[n+1];
		for(int i=1; i<=n; i++) {
			adj[i] = new ArrayList<Integer>();
		}
		for(int i=1; i<n; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			adj[a].add(b);
			adj[b].add(a);
		}
		dfs(1, -1);
		
		System.out.println(Math.max(dp[1][0], dp[1][1]));
	}
	static void dfs(int node, int parent) {
		dp[node][0] = 0;
		dp[node][1] = -M;
		for(int v: adj[node]) {
			if(v!=parent) {
				dfs(v, node);
				dp[node][0] += Math.max(dp[v][0], dp[v][1]);
				int temp = Math.min(dp[v][0] - dp[v][1], 0);
				dp[node][1] = Math.max(dp[node][1], temp);
			}
		}
		dp[node][1] += dp[node][0]+1;
	}
}

File: C++Projects/X_Camp/CS401H_TreeMatching.cpp
#include <bits/stdc++.h>
using namespace std;

const int M = 2*10e5+1;
int dp[M][2];
vector<int> adj[M];

void dfs(int node , int parent) {
    dp[node][0] = 0;
    dp[node][1] = -M;
    for(int v : adj[node]){
        if(v!=parent){
            dfs(v, node);
            dp[node][0] += max(dp[v][0], dp[v][1]);
            int temp = min(dp[v][0]-dp[v][1], 0);
            dp[node][1] = max(dp[node][1], temp);
        }
    }
    dp[node][1] += dp[node][0]+1;
}

int main(){
    int n; cin >> n;
    for(int i=1; i<n; i++){
        int a, b; cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    dfs(1, -1);
    cout << max(dp[1][0], dp[1][1]);
}